src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/AWS/SQS/CachedMessageHeadersHelper.cs

```diff
@@ -0,0 +1,74 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Reflection.Emit;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.SQS
+{
+    internal static class CachedMessageHeadersHelper<TMarkerType>
+    {
+        private const string StringDataType = "String";
+
+        private static readonly Func<string, object> _createMessageAttributeValue;
+        private static readonly Func<object> _createDict;
+
+        static CachedMessageHeadersHelper()
+        {
+            // Initialize delegate for creating a MessageAttributeValue object
+            var messageAttributeValueType = typeof(TMarkerType).Assembly.GetType("Amazon.SQS.Model.MessageAttributeValue");
+            var messageAttributeValueCtor = messageAttributeValueType.GetConstructor(System.Type.EmptyTypes);
+
+            DynamicMethod createMessageAttributeValueMethod = new DynamicMethod(
+                $"KafkaCachedMessageHeadersHelpers",
+                messageAttributeValueType,
+                parameterTypes: new Type[] { typeof(string) },
+                typeof(DuckType).Module,
+                true);
+
+            ILGenerator messageAttributeIL = createMessageAttributeValueMethod.GetILGenerator();
+            messageAttributeIL.Emit(OpCodes.Newobj, messageAttributeValueCtor);
+
+            messageAttributeIL.Emit(OpCodes.Dup);
+            messageAttributeIL.Emit(OpCodes.Ldstr, StringDataType);
+            messageAttributeIL.Emit(OpCodes.Callvirt, messageAttributeValueType.GetProperty("DataType").GetSetMethod());
+
+            messageAttributeIL.Emit(OpCodes.Dup);
+            messageAttributeIL.Emit(OpCodes.Ldarg_0);
+            messageAttributeIL.Emit(OpCodes.Callvirt, messageAttributeValueType.GetProperty("StringValue").GetSetMethod());
+
+            messageAttributeIL.Emit(OpCodes.Ret);
+
+            _createMessageAttributeValue = (Func<string, object>)createMessageAttributeValueMethod.CreateDelegate(typeof(Func<string, object>));
+
+            // Initialize delegate for creating a Dictionary<string, MessageAttributeValue> object
+            var genericDictType = typeof(Dictionary<,>);
+            var constructedDictType = genericDictType.MakeGenericType(new Type[] { typeof(string), messageAttributeValueType });
+            ConstructorInfo dictionaryCtor = constructedDictType.GetConstructor(System.Type.EmptyTypes);
+
+            DynamicMethod createDictMethod = new DynamicMethod(
+                $"KafkaCachedMessageHeadersHelpers",
+                constructedDictType,
+                null,
+                typeof(DuckType).Module,
+                true);
+
+            ILGenerator dictIL = createDictMethod.GetILGenerator();
+            dictIL.Emit(OpCodes.Newobj, dictionaryCtor);
+            dictIL.Emit(OpCodes.Ret);
+
+            _createDict = (Func<object>)createDictMethod.CreateDelegate(typeof(Func<object>));
```

Comment:
I'm guessing it doesn't make any difference, but could we make this delegate `Func<IDictionary>`, to save the cast in the method below? Just a little more explicit :)

Reply:
Yeah I can do that. Fixed in aec4cd2f86db5039a545fa176f147f7a609044ab